crates/levm/src/vm.rs

```diff
@@ -25,6 +26,54 @@ impl VM {
                     self.stack.push(value_to_push);
                     self.increment_pc_by(32);
                 }
+                Opcode::MLOAD => {
+                    // spend_gas(3);
+                    let offset = self.stack.pop().unwrap();
+                    // resize if necessary
+                    self.memory.resize(offset.as_usize());
+
+                    let value = self.memory.load(offset.as_usize());
+                    self.stack.push(value);
+                }
+                Opcode::MSTORE => {
+                    // spend_gas(3);
+                    let offset = self.stack.pop().unwrap().as_usize();
+                    let value = self.stack.pop().unwrap();
+                    let mut value_bytes = [0u8; 32];
+                    value.to_big_endian(&mut value_bytes);
+                    self.memory.resize(offset);
+
+                    self.memory.store_bytes(offset, &value_bytes);
+                }
+                Opcode::MSTORE8 => {
+                    // spend_gas(3);
+                    let offset = self.stack.pop().unwrap().as_usize();
+                    let value = self.stack.pop().unwrap();
+                    let mut value_bytes = [0u8; 32];
+                    value.to_big_endian(&mut value_bytes);
+                    self.memory.resize(offset);
+
+                    self.memory
+                        .store_bytes(offset, value_bytes[31..32].as_ref());
+                }
+                Opcode::MSIZE => {
+                    // spend_gas(2);
+                    self.stack.push(self.memory.size());
+                }
+                Opcode::MCOPY => {
+                    // spend_gas(3) + dynamic gas
+                    let dest_offset = self.stack.pop().unwrap().as_usize();
+                    let src_offset = self.stack.pop().unwrap().as_usize();
+                    let size = self.stack.pop().unwrap().as_usize();
+                    if size == 0 {
+                        continue;
+                    }
+
+                    let max_size = std::cmp::max(src_offset + size, dest_offset + size);
+                    self.memory.resize(max_size);
```

Comment:
I suggest moving the resize logic to the different methods since all of them do it anyways.

Reply:
done!